File: frontend/src/simulation/EnvironmentalField.js
/**
 * Environmental Field - Stores collective memory traces
 * The world itself becomes a memory storage medium
 */

export class EnvironmentalField {
  constructor(worldSize) {
    this.worldSize = worldSize;
    
    // Grid-based pheromone field
    this.gridSize = 20; // Cell size
    this.gridWidth = Math.ceil(worldSize.width / this.gridSize);
    this.gridHeight = Math.ceil(worldSize.height / this.gridSize);
    
    // Multiple pheromone layers
    this.pheromones = {
      danger: this.createGrid(),
      food: this.createGrid(),
      mating: this.createGrid(),
      death: this.createGrid(),
      activity: this.createGrid()
    };
    
    // Memory anchors - strong persistent memories
    this.memoryAnchors = [];
    
    // Decay rates for different pheromones
    this.decayRates = {
      danger: 0.99,
      food: 0.98,
      mating: 0.97,
      death: 0.995,
      activity: 0.96
    };
  }
  
  createGrid() {
    const grid = [];
    for (let y = 0; y < this.gridHeight; y++) {
      grid[y] = new Float32Array(this.gridWidth);
    }
    return grid;
  }
  
  // Deposit trace from organism
  depositTrace(trace) {
    const { position, pheromone, intensity = 1.0 } = trace;
    
    // Validate intensity
    const safeIntensity = Number.isFinite(intensity) ? Math.min(10, Math.max(0, intensity)) : 1.0;
    
    // Convert position to grid coordinates
    const gx = Math.floor(position.x / this.gridSize);
    const gy = Math.floor(position.y / this.gridSize);
    
    if (gx < 0 || gx >= this.gridWidth || gy < 0 || gy >= this.gridHeight) {
      return;
    }
    
    // Deposit pheromones with clamping to prevent overflow
    if (pheromone[0] > 0) {
      this.pheromones.danger[gy][gx] = Math.min(10, this.pheromones.danger[gy][gx] + pheromone[0] * safeIntensity);
    }
    if (pheromone[1] > 0) {
      this.pheromones.food[gy][gx] = Math.min(10, this.pheromones.food[gy][gx] + pheromone[1] * safeIntensity);
    }
    if (pheromone[2] > 0) {
      this.pheromones.mating[gy][gx] = Math.min(10, this.pheromones.mating[gy][gx] + pheromone[2] * safeIntensity);
    }
    if (pheromone[3] > 0) {
      this.pheromones.death[gy][gx] = Math.min(10, this.pheromones.death[gy][gx] + pheromone[3] * safeIntensity);
    }
    if (pheromone[4] > 0) {
      this.pheromones.activity[gy][gx] = Math.min(10, this.pheromones.activity[gy][gx] + pheromone[4] * safeIntensity);
    }
    
    // Diffuse to neighboring cells
    this.diffuse(gx, gy, safeIntensity * 0.3);
  }
  
  diffuse(cx, cy, amount) {
    // Simple diffusion to 8 neighbors
    for (let dy = -1; dy <= 1; dy++) {
      for (let dx = -1; dx <= 1; dx++) {
        if (dx === 0 && dy === 0) continue;
        
        const nx = cx + dx;
        const ny = cy + dy;
        
        if (nx >= 0 && nx < this.gridWidth && ny >= 0 && ny < this.gridHeight) {
          const distance = Math.sqrt(dx * dx + dy * dy);
          const diffuseAmount = amount / (distance * 2);
          
          for (const [type, grid] of Object.entries(this.pheromones)) {
            grid[ny][nx] += grid[cy][cx] * diffuseAmount;
          }
        }
      }
    }
  }
  
  // Add memory anchor for significant events
  addMemoryAnchor(anchor) {
    this.memoryAnchors.push({
      ...anchor,
      timestamp: Date.now(),
      strength: 1.0
    });
    
    // Limit anchors
    if (this.memoryAnchors.length > 100) {
      // Remove weakest anchors
      this.memoryAnchors.sort((a, b) => b.strength - a.strength);
      this.memoryAnchors = this.memoryAnchors.slice(0, 100);
    }
  }
  
  // Read environmental memory at position
  readEnvironment(position, radius = 50) {
    const gx = Math.floor(position.x / this.gridSize);
    const gy = Math.floor(position.y / this.gridSize);
    
    const memory = {
      danger: 0,
      food: 0,
      mating: 0,
      death: 0,
      activity: 0,
      nearbyAnchors: []
    };
    
    // Sample pheromones in radius
    const gridRadius = Math.ceil(radius / this.gridSize);
    let count = 0;
    
    for (let dy = -gridRadius; dy <= gridRadius; dy++) {
      for (let dx = -gridRadius; dx <= gridRadius; dx++) {
        const nx = gx + dx;
        const ny = gy + dy;
        
        if (nx >= 0 && nx < this.gridWidth && ny >= 0 && ny < this.gridHeight) {
          const dist = Math.sqrt(dx * dx + dy * dy) * this.gridSize;
          if (dist <= radius) {
            const weight = 1 - (dist / radius);
            
            memory.danger += this.pheromones.danger[ny][nx] * weight;
            memory.food += this.pheromones.food[ny][nx] * weight;
            memory.mating += this.pheromones.mating[ny][nx] * weight;
            memory.death += this.pheromones.death[ny][nx] * weight;
            memory.activity += this.pheromones.activity[ny][nx] * weight;
            
            count++;
          }
        }
      }
    }
    
    // Normalize
    if (count > 0) {
      memory.danger /= count;
      memory.food /= count;
      memory.mating /= count;
      memory.death /= count;
      memory.activity /= count;
    }
    
    // Find nearby anchors
    memory.nearbyAnchors = this.memoryAnchors.filter(anchor => {
      const dx = anchor.position.x - position.x;
      const dy = anchor.position.y - position.y;
      return Math.sqrt(dx * dx + dy * dy) < radius;
    });
    
    return memory;
  }
  
  // Update field over time
  update(deltaTime) {
    // Decay pheromones
    for (const [type, grid] of Object.entries(this.pheromones)) {
      const decay = Math.pow(this.decayRates[type], deltaTime);
      
      for (let y = 0; y < this.gridHeight; y++) {
        for (let x = 0; x < this.gridWidth; x++) {
          grid[y][x] *= decay;
          
          // Remove very small values
          if (grid[y][x] < 0.001) {
            grid[y][x] = 0;
          }
        }
      }
    }
    
    // Decay memory anchors
    this.memoryAnchors = this.memoryAnchors.filter(anchor => {
      anchor.strength *= Math.pow(0.999, deltaTime);
      return anchor.strength > 0.01;
    });
  }
  
  // Get visualization data
  getVisualizationData() {
    const data = [];
    
    for (let y = 0; y < this.gridHeight; y++) {
      for (let x = 0; x < this.gridWidth; x++) {
        const danger = this.pheromones.danger[y][x];
        const food = this.pheromones.food[y][x];
        const activity = this.pheromones.activity[y][x];
        
        // Validate values are finite
        const safeDanger = Number.isFinite(danger) ? Math.min(10, Math.max(0, danger)) : 0;
        const safeFood = Number.isFinite(food) ? Math.min(10, Math.max(0, food)) : 0;
        const safeActivity = Number.isFinite(activity) ? Math.min(10, Math.max(0, activity)) : 0;
        
        if (safeDanger > 0.1 || safeFood > 0.1 || safeActivity > 0.1) {
          data.push({
            x: x * this.gridSize + this.gridSize / 2,
            y: y * this.gridSize + this.gridSize / 2,
            danger: safeDanger,
            food: safeFood,
            activity: safeActivity,
            size: this.gridSize
          });
        }
      }
    }
    
    return data;
  }
}

File: backend/proteus/environment/environmental_field.py
#!/usr/bin/env python3
"""
Environmental Field - Collective Memory System
Implements the third layer of PROTEUS inheritance
"""

import numpy as np
from numba import jit, prange
from typing import Dict, List, Tuple, Optional
import json


class EnvironmentalField:
    """
    Environmental memory field that organisms can read and write to
    Represents collective knowledge and experience trails
    """
    
    def __init__(self, world_size: Tuple[int, int], resolution: int = 20):
        self.world_size = world_size
        self.resolution = resolution
        
        # Grid dimensions
        self.grid_width = world_size[0] // resolution
        self.grid_height = world_size[1] // resolution
        
        # Pheromone layers (float32 for GPU compatibility)
        self.pheromones = {
            'danger': np.zeros((self.grid_height, self.grid_width), dtype=np.float32),
            'food': np.zeros((self.grid_height, self.grid_width), dtype=np.float32),
            'mating': np.zeros((self.grid_height, self.grid_width), dtype=np.float32),
            'death': np.zeros((self.grid_height, self.grid_width), dtype=np.float32),
            'activity': np.zeros((self.grid_height, self.grid_width), dtype=np.float32)
        }
        
        # Memory anchors - strong persistent memories
        self.memory_anchors = []
        
        # Decay rates
        self.decay_rates = {
            'danger': 0.98,
            'food': 0.99,
            'mating': 0.97,
            'death': 0.995,  # Death memories persist longer
            'activity': 0.96
        }
        
    @staticmethod
    @jit(nopython=True, parallel=True)
    def _deposit_pheromone_jit(grid, x, y, intensity, radius):
        """JIT-compiled pheromone deposition with diffusion"""
        h, w = grid.shape
        
        # Grid coordinates
        gx = int(x)
        gy = int(y)
        
        # Deposit with gaussian falloff
        for dy in prange(-radius, radius + 1):
            for dx in range(-radius, radius + 1):
                ny = gy + dy
                nx = gx + dx
                
                if 0 <= ny < h and 0 <= nx < w:
                    dist = np.sqrt(dx*dx + dy*dy)
                    if dist <= radius:
                        falloff = np.exp(-dist*dist / (radius*radius * 0.5))
                        grid[ny, nx] += intensity * falloff
                        
    def deposit_pheromone(self, position: Tuple[float, float], 
                         pheromone_type: str, intensity: float):
        """Deposit pheromone at a world position"""
        if pheromone_type not in self.pheromones:
            return
            
        # Convert world position to grid
        gx = position[0] / self.resolution
        gy = position[1] / self.resolution
        
        # Deposit with diffusion
        self._deposit_pheromone_jit(
            self.pheromones[pheromone_type],
            gx, gy, intensity, 2
        )
        
    def deposit_trace(self, organism_inheritance, position: Tuple[float, float], 
                     event_type: str, intensity: float = 1.0):
        """Deposit an environmental trace from an organism"""
        # Get organism's pheromone signature
        pheromone = organism_inheritance.pheromone_signature
        
        # Map event types to pheromone layers
        if event_type == 'death':
            self.deposit_pheromone(position, 'death', intensity * 2.0)
            self.deposit_pheromone(position, 'danger', intensity)
            # Create memory anchor for significant deaths
            if intensity > 1.5:
                self.add_memory_anchor(position, 'death', intensity)
                
        elif event_type == 'predator_escape':
            self.deposit_pheromone(position, 'danger', intensity * 1.5)
            
        elif event_type == 'feeding':
            self.deposit_pheromone(position, 'food', intensity)
            
        elif event_type == 'reproduction':
            self.deposit_pheromone(position, 'mating', intensity)
            
        # Always deposit activity
        self.deposit_pheromone(position, 'activity', intensity * 0.5)
        
    def add_memory_anchor(self, position: Tuple[float, float], 
                         anchor_type: str, strength: float):
        """Add a strong memory point that persists longer"""
        anchor = {
            'position': position,
            'type': anchor_type,
            'strength': strength,
            'timestamp': 0,  # Will be set by simulation
            'decay': 0.999  # Very slow decay
        }
        
        self.memory_anchors.append(anchor)
        
        # Limit anchors
        if len(self.memory_anchors) > 100:
            # Remove weakest
            self.memory_anchors.sort(key=lambda a: a['strength'])
            self.memory_anchors.pop(0)
            
    @staticmethod
    @jit(nopython=True)
    def _read_field_jit(grid, gx, gy, radius):
        """JIT-compiled field reading"""
        h, w = grid.shape
        total = 0.0
        count = 0
        
        for dy in range(-radius, radius + 1):
            for dx in range(-radius, radius + 1):
                ny = gy + dy
                nx = gx + dx
                
                if 0 <= ny < h and 0 <= nx < w:
                    dist = np.sqrt(dx*dx + dy*dy)
                    if dist <= radius:
                        weight = 1.0 / (1.0 + dist)
                        total += grid[ny, nx] * weight
                        count += 1
                        
        return total / max(count, 1)
        
    def read_field(self, position: Tuple[float, float], 
                   radius: float = 30) -> Dict[str, float]:
        """Read pheromone concentrations at a position"""
        # Convert to grid coordinates
        gx = int(position[0] / self.resolution)
        gy = int(position[1] / self.resolution)
        grid_radius = int(radius / self.resolution)
        
        readings = {}
        for ptype, grid in self.pheromones.items():
            readings[ptype] = self._read_field_jit(grid, gx, gy, grid_radius)
            
        return readings
        
    def get_gradient(self, position: Tuple[float, float], 
                    pheromone_type: str) -> Tuple[float, float]:
        """Get gradient direction of a pheromone type"""
        if pheromone_type not in self.pheromones:
            return (0.0, 0.0)
            
        # Sample around position
        delta = 10  # Sample distance
        center = self.read_field(position)[pheromone_type]
        
        # Calculate gradients
        right = self.read_field((position[0] + delta, position[1]))[pheromone_type]
        left = self.read_field((position[0] - delta, position[1]))[pheromone_type]
        up = self.read_field((position[0], position[1] - delta))[pheromone_type]
        down = self.read_field((position[0], position[1] + delta))[pheromone_type]
        
        gx = (right - left) / (2 * delta)
        gy = (down - up) / (2 * delta)
        
        return (gx, gy)
        
    @staticmethod
    @jit(nopython=True, parallel=True)
    def _update_field_jit(grid, decay_rate):
        """JIT-compiled field update with decay and diffusion"""
        h, w = grid.shape
        
        # Decay
        for i in prange(h):
            for j in range(w):
                grid[i, j] *= decay_rate
                
                # Clear very small values
                if grid[i, j] < 0.001:
                    grid[i, j] = 0.0
                    
    def update(self, dt: float):
        """Update the environmental field"""
        # Update each pheromone layer
        for ptype, grid in self.pheromones.items():
            self._update_field_jit(grid, self.decay_rates[ptype])
            
        # Update memory anchors
        for anchor in self.memory_anchors:
            anchor['strength'] *= anchor['decay']
            
        # Remove weak anchors
        self.memory_anchors = [a for a in self.memory_anchors if a['strength'] > 0.1]
        
    def get_visualization_data(self) -> List[Dict]:
        """Get data for visualization"""
        cells = []
        
        for y in range(self.grid_height):
            for x in range(self.grid_width):
                # Only include cells with significant activity
                danger = self.pheromones['danger'][y, x]
                food = self.pheromones['food'][y, x]
                activity = self.pheromones['activity'][y, x]
                
                if danger > 0.1 or food > 0.1 or activity > 0.1:
                    cells.append({
                        'x': x * self.resolution + self.resolution / 2,
                        'y': y * self.resolution + self.resolution / 2,
                        'danger': float(danger),
                        'food': float(food),
                        'activity': float(activity),
                        'size': self.resolution
                    })
                    
        return cells
        
    def save_state(self) -> Dict:
        """Save field state for persistence"""
        return {
            'pheromones': {k: v.tolist() for k, v in self.pheromones.items()},
            'memory_anchors': self.memory_anchors,
            'world_size': self.world_size,
            'resolution': self.resolution
        }
        
    def load_state(self, state: Dict):
        """Load field state"""
        for ptype, data in state['pheromones'].items():
            self.pheromones[ptype] = np.array(data, dtype=np.float32)
            
        self.memory_anchors = state['memory_anchors']